Add gamepad gyro and flick stick support

ceski-1 opened the pull request:
Work in progress. Fixes https://github.com/fabiangreffrath/woof/issues/1771

Changes:
- Added gamepad [flick stick](https://en.wikipedia.org/wiki/Flick_Stick) support
- Improvements to extra turn/look sensitivity
- Simplified gamepad menu
- Renamed gamepad settings
- New gamepad default settings to match most source ports
- Lots of cosmetic changes

To do:
- Gamepad gyro support
- Gamepad touchpad support?

rfomin commented:
Does the flick stick only work well with the gyroscope? I tried it on an old XBox 360 gamepad, it feels weird.
Otherwise I like the new default settings and the simplified menu.

ceski-1 commented:
> Edit: I think I get it now. I probably never get used to it, but oh well 😄

It's bizarre at first but felt okay over the course of testing. It needs to be combined with gyro to feel complete though. A bunch of competitive fps games have added this, as well as Steam, so it seems there is a small but hardcore audience that likes it.

ceski-1 commented:
Changes:
- Added gyro support. Based on the best known implementation by Jibb Smart.
- Added float type for events to support SDL gyro data. This will also be useful in the future for SDL3 mouse input.

To do:
- Not sure what to do with placeholder gyro menu yet. Add a submenu under "Gamepad" tab? Reorganize "Mouse/Gamepad/Gyro" under one "Input" tab?
- Manual calibration support (gyros drift):
  - Add a menu item that is just "Calibrate..." and pops up a box to show progress (similar to reset defaults).
  - Save calibration data (4 floats) to woof.cfg. Maybe save as strings and load with `strtof`?
- Simplify vector functions.
- Work out sensible defaults.
- Documentation.

![woof0003](https://github.com/user-attachments/assets/bf1aeb1c-36ea-4ab6-b75c-bd30e70019d6)

fabiangreffrath commented:
> * Added gyro support. Based on the best known implementation by Jibb Smart.

Does your implementation involve code from Jibb Smart?

ceski-1 commented:
> Does your implementation involve code from Jibb Smart?

I wrote the implementation based on the five guides listed here with sample code:
http://gyrowiki.jibbsmart.com

His wiki is CC BY 4.0. All code can be used freely, just include a link back (I already added them in the code comments):
http://gyrowiki.jibbsmart.com/legal:_home

He also has a project that uses that similar code, under the MIT license, but I couldn't follow it vey well, so I went back to the wiki:
https://github.com/Electronicks/JoyShockMapper

Should we do more to attribute this? README.md additions?

fabiangreffrath commented:
If you have taken *code* from him, it needs to carry his copyright and license. If you have merely taken *ideas*, it's okay to mention this in the code comments. Either way, we should of course mention gyro controls in the README file and we can take the opportunity to attribute this to his work. Just realizing we should have done the same with Andrew Apted and his voxel ad nanobsp code.

ceski-1 commented:
Okay, added readme changes here: https://github.com/fabiangreffrath/woof/pull/1800/commits/aa0e8e6852e03c9d63804c5ccf30acdda0d0f2df

fabiangreffrath commented:
> Okay, added readme changes here: [aa0e8e6](https://github.com/fabiangreffrath/woof/commit/aa0e8e6852e03c9d63804c5ccf30acdda0d0f2df)

Thank you! But now, this contradicts the source file headers. Please add your own copyright in README.md and Jibb Smart's copyright in the affected file and change the license to CC-BY-4.0, if you agree, or consider dual-licensing your work.

rfomin commented:
Impressive work! I don't have a gamepad with a gyro, so I can't test it or comment on the options.

> * Not sure what to do with placeholder gyro menu yet. Add a submenu under "Gamepad" tab? Reorganize "Mouse/Gamepad/Gyro" under one "Input" tab?

I think it's okay to have a "Gyro" tab.

> * Manual calibration support (gyros drift)

I guess we need a submenu for that. How is this done in other games?

ceski-1 commented:
> I think it's okay to have a "Gyro" tab.

Another concept:

```
   item name  on
   some item  off
        item  [--|------]

   > submenu  ... <
  submenu #2  ...
```

In this concept, the submenu is formatted the same as the regular one. You can return to the first menu by clicking the tab at the top, or a different tab, or pressing backspace/padB. Maybe too complicated.

<details><summary>Rough example:</summary>

![woof0001](https://github.com/user-attachments/assets/1ddd1034-6e1a-4301-9d5e-fe3f976f4b9f)

</details>


> I guess we need a submenu for that. How is this done in other games?

This is how it works in yquake2 (yamagi): You set the controller down, then click "Calibrate". A box pops up that looks just like the "reset to defaults" black rectangle. The text says "Calibrating..." then after collecting samples for about 1-2 seconds the text changes to "Caibrating Done!" and closes automatically. Input should be probably be blocked for those 1-2 seconds.

fabiangreffrath commented:
> Rough example:

I like the "Gyro Options >>>" concept!

> This is how it works in yquake2 (yamagi): You set the controller down, then click "Calibrate". A box pops up that looks just like the "reset to defaults" black rectangle. The text says "Calibrating..." then after collecting samples for about 1-2 seconds the text changes to "Caibrating Done!" and closes automatically. Input should be probably be blocked for those 1-2 seconds.

Sounds reasonable.

ceski-1 commented:
Here is how Ironwail does it. I like it:


https://github.com/user-attachments/assets/9196a68a-e69b-4630-af6f-e32b393b4d07

ceski-1 commented on src/mn_internal.h:
```diff
@@ -102,7 +102,7 @@ void MN_DrawEnemy(void);
 #define S_SELECT      0x00000002 // We're changing this item
 #define S_TITLE       0x00000004 // Title item
-
+#define S_FUNC        0x00000008 // Call var.func() on MENU_ENTER
```
This uses one of the only flags left here.

rfomin replied:
No problem, we can add `flags2` or make `flags` 64-bit.

ceski-1 commented:
https://github.com/user-attachments/assets/887ee383-54ef-4243-b142-a5c73de85431

ceski-1 commented:
I think I'll stop here. I would like to have a submenu for gyro, but it looks like it would require reworking of the menu system, so maybe that's for another time.

rfomin commented:
> I would like to have a submenu for gyro, but it looks like it would require reworking of the menu system, so maybe that's for another time.

Sorry, I just haven't done anything in the last two mounths. I don't think submenu is complicated, it kinda works in #1743. But maybe we should add scrolling to the menus? Also may be useful in level table.

ceski-1 commented:
This is similar to a feature in Ironwail but is an original implementation. Raw data from the gamepad is displayed as a dot to help tune ideal deadzones and gyro steadying.

https://github.com/user-attachments/assets/3e211a79-264e-4b4c-bbb8-2d86db35f56c

fabiangreffrath commented:
:wow:

rfomin commented:
Wow, looks cool!

fabiangreffrath reviewed (approved):
This code change is a bit diluted with semi-unrelated changes like the "controller" -> "gamepad" rename or the introduction of the `union evdata_t` type. However, code-wise it looks clean and comprehensible as always. Thank you very much!

I'll try to get my son's switch controller connected tomorrow and provide some feedback.

ceski-1 commented:
Last push was just these two: https://github.com/fabiangreffrath/woof/pull/1800/commits/e5a0bf8b0c5afef289bf5593d7b6f63bc58db1ed and https://github.com/fabiangreffrath/woof/pull/1800/commits/8bb76dd089d496c9c3eb030361624ab69adaac66

ceski-1 commented on src/mn_setup.c:
```diff
@@ -1074,7 +1091,9 @@ static void DrawInstructions()
         {
             if (menu_input == pad_mode)
             {
-                s = "[ PadA ] to toggle";
+                M_snprintf(pad_buf, sizeof(pad_buf), "[ %s ] to toggle",
+                           str_pad[STR_PADA]);
+                s = pad_buf;
```
Is `M_snprintf` the best way to go here? I'm always wary of strings in C.

fabiangreffrath replied:
Yes, when writing a formatted string to a fixed size buffer, I'd say `M_snprintf()` is the preferred method.

fabiangreffrath commented on src/i_gamepad.h:
```diff
@@ -27,6 +27,14 @@
 #define RAD2TIC(x) ((x) * 32768.0f / PI_F) // Radians to ticcmd angle.
 #define LENGTH_F(x, y) sqrtf((x) * (x) + (y) * (y))
 
+typedef enum joy_platform_e
```
This is cool! Could we get something similar for the Nintendo Switch and its Y/X/B/A button names? Or does this already match with the XBox layout?

ceski-1 replied:
Sure, I can add Switch naming. The current differences are minor (e.g. back/start in Woof instead of minus/plus on a Switch controller). You may have noticed that Switch's A/B and X/Y are in different positions than the common standard, but it's still A as confirm and B as cancel. That's just how Nintendo does things.

fabiangreffrath replied:
Yes, right. But I guess it would still help if we had presets for the three most common consoles out there.

ceski-1 replied:
Fixed by https://github.com/fabiangreffrath/woof/pull/1800/commits/8552e0ec5ee015895e338378813500306286d58d

ceski-1 commented:
There are some gyro issues with Switch controllers which is going to take some time to fix.

ceski-1 commented:
> There are some gyro issues with Switch controllers which is going to take some time to fix.

Fixed by https://github.com/fabiangreffrath/woof/pull/1800/commits/8916eb726768fb5e16b4a664f7842ac8fda4eaae
